File: PY/insertion_sort.py
import random
import time
import sys


def insertion_sort(arr):  

    for step in range(1, len(arr)):
        key = arr[step]
        j = step - 1

        while j >= 0 and key < arr[j]:
            arr[j + 1] = arr[j]
            j = j - 1

        arr[j + 1] = key


def generate_random_array(length):
    array = list(range(1, length + 1))  
    random.shuffle(array)  
    return array

def compare_arrays(array1, array2):
    if len(array1) != len(array2):
        return False

    for i in range(len(array1)):
        if array1[i] != array2[i]:
            return False

    return True

def test(n):

    n = int(n);    

    random_array = generate_random_array(n)

    correct_array = list(range(1, n + 1)) 

    start_time = time.time()
    insertion_sort(random_array)
    end_time = time.time()
    execution_time = end_time - start_time


    print(compare_arrays(random_array, correct_array)) 
    print(f"Function execution time: {execution_time:.6f} seconds")

if len(sys.argv) < 2:
    print("Usage: python script.py <argument>")
    sys.exit(1)

script_name = sys.argv[0]
argument = sys.argv[1]

test(argument)

print("Script name:", script_name)
print("Argument:", argument)

File: C/insertionSort.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <time.h>

void insertionSort(int* arr, int n) {
    for (int i = 1; i < n; i++) {
        int key = arr[i];
        int j = i - 1;

        while (j >= 0 && arr[j] > key) {
            arr[j + 1] = arr[j];
            j--;
        }

        arr[j + 1] = key;
    }   
}

int* generateArray(int n) 
{

    int* arr = (int*)malloc(n * sizeof(int)); 

    if (arr == NULL) 
    {
        perror("Memory allocation failed");
        exit(1); 
    }

    for (int i = 0; i < n; i++) 
    {
        arr[i] = i + 1;
    }

    return arr;
}

int* generateRandomArray(int n) 
{

    int* arr = generateArray(n);

    srand(time(NULL));

    for (int i = n - 1; i > 0; i--) 
    {
        int j = rand() % (i + 1);
        int temp = arr[i];
        arr[i] = arr[j];
        arr[j] = temp;
    }

    return arr;
}

bool compareArrays(int arr1[], int arr2[], int size) 
{
    for (int i = 0; i < size; i++) 
    {
        if (arr1[i] != arr2[i]) {
            return false; 
        }
    }
    return true; 
}

int main(int argc, char *argv[]) {
    
    int n = atoi(argv[1]);

    int* correctArray = generateArray(n);

    int* randomArray = generateRandomArray(n);

    clock_t start_time, end_time;
    double cpu_time_used;

    start_time = clock();
    insertionSort(randomArray, n);
    end_time = clock();
    cpu_time_used = ((double)(end_time - start_time)) / CLOCKS_PER_SEC;

    bool isCorrect = compareArrays(randomArray, correctArray, n);

    free(correctArray);
    free(randomArray);

    if (isCorrect) {
        printf("Correct, boi\n");
    } else {
        printf("Wrong, boi\n");
    }

    printf("Execution Time: %f seconds\n", cpu_time_used);
}
